Makefile

```diff
@@ -152,27 +149,30 @@ ELF := $(ROM:.z64=.elf)
 # description of ROM segments
 SPEC := spec
 
-ifeq ($(COMPILER),ido)
-SRC_DIRS := $(shell find src -type d -not -path src/gcc_fix)
-else
-SRC_DIRS := $(shell find src -type d)
+# Folders ZAPD extracts assets to
+ASSETS_ZAPD_DEST_DIRS := assets/code assets/misc assets/objects assets/overlays assets/scenes assets/textures
+# Create the folders to avoid find printing errors
+stdout := $(shell mkdir -p $(ASSETS_ZAPD_DEST_DIRS))
+ifneq ($(stdout),)
+$(error $(stdout))
 endif
 
-ASSET_BIN_DIRS := $(shell find assets/* -type d -not -path "assets/xml*" -not -path "assets/text")
-ASSET_FILES_XML := $(foreach dir,$(ASSET_BIN_DIRS),$(wildcard $(dir)/*.xml))
-ASSET_FILES_BIN := $(foreach dir,$(ASSET_BIN_DIRS),$(wildcard $(dir)/*.bin))
-ASSET_FILES_OUT := $(foreach f,$(ASSET_FILES_XML:.xml=.c),$f) \
-				   $(foreach f,$(ASSET_FILES_BIN:.bin=.bin.inc.c),build/$f) \
-				   $(foreach f,$(wildcard assets/text/*.c),build/$(f:.c=.o))
+# Folders under which source files are to be found
+SRC_DIRS := assets/text src/boot src/buffers src/code src/dmadata src/elf_message src/libultra src/makerom src/overlays
+ifeq ($(COMPILER),gcc)
+SRC_DIRS += src/gcc_fix
+endif
 
-UNDECOMPILED_DATA_DIRS := $(shell find data -type d)
+ASSETS_ZAPD_FILES_BIN := $(shell find $(ASSETS_ZAPD_DEST_DIRS) -name '*.bin')
+ASSETS_ZAPD_FILES_PNG := $(shell find $(ASSETS_ZAPD_DEST_DIRS) -name '*.png')
+ASSETS_ZAPD_FILES_JPG := $(shell find $(ASSETS_ZAPD_DEST_DIRS) -name '*.jpg')
+ASSETS_ZAPD_FILES_OUT := $(addprefix build/,$(ASSETS_ZAPD_FILES_BIN:.bin=.bin.inc.c) $(ASSETS_ZAPD_FILES_PNG:.png=.inc.c) $(ASSETS_ZAPD_FILES_JPG:.jpg=.jpg.inc.c))
 
-# source files
-C_FILES       := $(filter-out %.inc.c,$(foreach dir,$(SRC_DIRS) $(ASSET_BIN_DIRS),$(wildcard $(dir)/*.c)))
-S_FILES       := $(foreach dir,$(SRC_DIRS) $(UNDECOMPILED_DATA_DIRS),$(wildcard $(dir)/*.s))
-O_FILES       := $(foreach f,$(S_FILES:.s=.o),build/$f) \
-                 $(foreach f,$(C_FILES:.c=.o),build/$f) \
-                 $(foreach f,$(wildcard baserom/*),build/$f.o)
+# Source files
+# List all .c files in assets and source directories, excluding .inc.c files
+C_FILES := $(shell find $(ASSETS_ZAPD_DEST_DIRS) $(SRC_DIRS) -name '*.inc.c' -o -name '*.c' -print)
+S_FILES := $(shell find $(SRC_DIRS) data -name '*.s')
```

Comment:
Why replace `foreach` with `shell find` here?

Reply:
I changed `SRC_DIRS` to only list the "root" folders

Reply:
we have to maintain it manually anyway, because of conditional "inclusion" of src/gcc_fix, and it contains `assets/text` which isn't in src/

Reply:
We don't currently have to manually maintain every inclusion of every source dir on account of `src/gcc_fix`, since we use `find` to create the list, so I don't see what this change achieves at least with respect to the source dirs?

Reply:
basically I think rather than (master) "the ido path knows about gcc_fix and removes it"
```mk
ifeq ($(COMPILER),ido)
SRC_DIRS := $(shell find src -type d -not -path src/gcc_fix)
else
SRC_DIRS := $(shell find src -type d)
endif
```
it's cleaner to do (this PR) "add gcc_fix if relevant"
```mk
SRC_DIRS := assets/text src/audio src/boot src/buffers src/code src/dmadata src/elf_message src/libultra src/makerom src/overlays
ifeq ($(COMPILER),gcc)
SRC_DIRS += src/gcc_fix
endif
```
and it allows moving in assets/text there too as a bonus

Reply:
Note I would agree there is an underlying issue here of poor structure of the repo

Reply:
Is this possible?
```make
SRC_DIRS := assets/text $(shell find src -type d -not -path src/gcc_fix)
ifeq ($(COMPILER),gcc)
SRC_DIRS += src/gcc_fix
endif
```

Reply:
Why do we conditionally compile the files from `gcc_fix` instead of just building them and ignoring them on IDO builds?
If it is because of the `__asm__` statement then we could either move that asm into its own `.s` file or wrap it into a macro that expands to nothing on IDO.
If the C functions are problematic then we could just ifdef the whole file, so building with IDO just produces an empty file.

Reply:
Wrapping the file in `#ifndef __sgi` seems like a good option